src/modular/const_monty_form/macros.rs

```diff
@@ -21,27 +21,20 @@ macro_rules! impl_modulus {
             $uint_type: $crate::ConcatMixed<MixedOutput = $crate::Uint<DLIMBS>>,
         {
             const LIMBS: usize = <$uint_type>::LIMBS;
-            const MODULUS: $crate::NonZero<$uint_type> = {
+            const MODULUS: $crate::Odd<$uint_type> = {
                 let res = <$uint_type>::from_be_hex($value);
-
-                // Check that the modulus is odd
-                if res.as_limbs()[0].0 & 1 == 0 {
-                    panic!("modulus must be odd");
-                }
-
-                // Can unwrap here since `res` was asserted to be odd.
-                res.to_nz().expect("modulus ensured non-zero")
+                res.to_odd().expect("modulus must be odd")
```

Comment:
I think the phrasing in "expect" messages is supposed to be in line with the verb, that is `.expect("modulus ensured odd")`. That said, the formatting in `panic!()` calls in `ConstCtOption` may need to be adjusted (e.g. `assert!(self.is_some.is_true_vartime(), "Expected {}", msg);`)

Reply:
That's not what the documentation says: https://doc.rust-lang.org/std/option/enum.Option.html#recommended-message-style

They explicitly suggest phrasing like "slice should not be empty" (although as someone who constantly reads documents with RFC2119 phrasing, I prefer "must" over "should" for cases like this)

Reply:
I personally think it's a bad recommendation, and in that example `.expect("non-empty slice");` reads much better. But if we are to follow it, a lot of `expect()` usages in the crate should be amended.

Reply:
I think the "ensured" phrasing is fine in cases where code earlier in the function actually ensured something, i.e. internal usages, such as the ones around the existence of modular inverses in these constructors.

For external usages that might hit end-users, I find the kind of wording recommended in `Option::expect` much less confusing.

Reply:
Mostly because the corresponding `panic()` calls in `expect()` implementations expect the recommended phrasing from the docs. That is, they just print the message verbatim, without an "expects: " prefix, so what reads good in the code (with `expect` present) is confusing for the user and vice versa.

Reply:
I think "modulus must be odd" is quite clear in those kinds of cases, whereas "modulus ensured non-zero" is not

Reply:
It is different though. "modulus must be odd" means that the modulus may not be odd at this point, in which case there's a panic. "modulus ensured odd" means that the modulus *cannot* be odd at this point due to the previous code, and there won't be a panic (unless there's a bug in the code).

Reply:
Yeah, that's what I just said re: internal usages, highlighting `inv_mod2k_vartime` (though there I used the wording "modular inverse should exist").

But this particular instance is for an external-facing user error, where they have supplied an even modulus to `impl_modulus!` by mistake.

Reply:
But the reason for the user facing/dev facing difference is just the formatting of the message in the implementation of `expect()`. We can change the formatting in `ConstCtOption` thus eliminating that difference.

Reply:
Ah wait, I see what you're saying. This is another thing, `expect` being used as "try to unwrap with a message on failure" or as "assert that this should always be true unless there are bugs". These should really be two different methods, but we have what we have.

Reply:
Yeah, the message is particularly relevant for the compile error surfaced for an invalid modulus:

```
error[E0080]: evaluation of `<Modulus as ConstMontyParams<{ <$uint_type>::LIMBS }>>::MODULUS` failed
   --> /Users/tarcieri/src/RustCrypto/crypto-bigint/src/const_choice.rs:317:9
    |
317 |         assert!(self.is_some.is_true_vartime(), "{}", msg);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'modulus must be odd', /Users/tarcieri/src/RustCrypto/crypto-bigint/src/const_choice.rs:317:9
```

Perhaps the error would be better as an explicit (const) panic instead of using `expect`, because it's a little bit hard to tell what's going on there.

Reply:
I removed the `expect` here. It's now just:

```rust
            const MODULUS: $crate::Odd<$uint_type> = $crate::Odd::<$uint_type>::from_be_hex($value);
```

The `from_be_hex` method calls the same named method on `Uint` and does an assert that the returned value is odd or panics: https://github.com/RustCrypto/crypto-bigint/pull/487/files#diff-747dcf65a56c97dcb57a534812fd0eaa55a56a8c3f5da3d0267456deb3a10cd9R49

This makes for a better error message:

```
error[E0080]: evaluation of `<TestMod as ConstMontyParams<{ <$uint_type>::LIMBS }>>::MODULUS` failed
  --> /Users/tarcieri/src/RustCrypto/crypto-bigint/src/odd.rs:49:9
   |
49 |         assert!(uint.is_odd().is_true_vartime(), "number must be odd");
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'number must be odd', /Users/tarcieri/src/RustCrypto/crypto-bigint/src/odd.rs:49:9
```